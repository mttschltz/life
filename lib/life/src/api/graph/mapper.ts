import {
  Risk as RiskUsecase,
  Category as CategoryUsecase,
  CategoryTopLevel as CategoryTopLevelUsecase,
  Updated as UpdatedUsecase,
  isUpdatedCategory as isUpdatedCategoryUsecase,
  isUpdatedRisk as isUpdatedRiskUsecase,
  Impact as ImpactUsecase,
  Likelihood as LikelihoodUsecease,
  RiskType as RiskTypeUsecase,
} from '@life/usecase/mapper'
import { Category, CategoryTopLevel, Impact, Likelihood, Risk, RiskType } from '@life/__generated__/graphql'
import { Result, resultError, resultOk, results, Results } from '@helper/result'

type MdxTranspiler = (mdx?: string) => string | undefined

interface GraphMapper {
  categoryTopLevelToUsecase: (graphCategory: CategoryTopLevel) => Result<CategoryTopLevelUsecase>
  riskFromUsecase: (risk: RiskUsecase) => Result<Risk>
  risksFromUsecase: (risks: RiskUsecase[]) => Results<Risk>
  updatedFromUsecase: (updated: UpdatedUsecase[]) => Results<Category | Risk>
  categoryFromUsecase: (category: CategoryUsecase) => Category
  categoriesFromUsecase: (categories: CategoryUsecase[]) => Category[]
  isUpdatedCategory: (u: Category | Risk) => u is Category
  isUpdatedRisk: (u: Category | Risk) => u is Risk
}

function newMapper(mdxTranspiler: MdxTranspiler): GraphMapper {
  return new GraphMapperImpl(mdxTranspiler)
}

class GraphMapperImpl implements GraphMapper {
  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  #mdxTranspiler: MdxTranspiler

  public constructor(mdxTranspiler: MdxTranspiler) {
    this.#mdxTranspiler = mdxTranspiler
  }

  public categoryFromUsecase(category: CategoryUsecase): Category {
    const parent = category.parent ? this.categoryFromUsecase(category.parent) : undefined
    const children = category.children.map((c) => this.categoryFromUsecase(c))
    return {
      id: category.id,
      name: category.name,
      path: category.path,
      description: category.description ?? undefined,
      shortDescription: category.shortDescription,
      parent,
      children,
      updated: category.updated,
    }
  }

  public categoriesFromUsecase(categories: CategoryUsecase[]): Category[] {
    return categories.map((c) => this.categoryFromUsecase(c))
  }

  // ignore code coverage for risks until they are refactored
  /* istanbul ignore next */
  public categoryTopLevelToUsecase(graphCategory: CategoryTopLevel): Result<CategoryTopLevelUsecase> {
    switch (graphCategory) {
      case CategoryTopLevel.Health:
        return resultOk<CategoryTopLevelUsecase>('Health')
      case CategoryTopLevel.Wealth:
        return resultOk<CategoryTopLevelUsecase>('Wealth')
      case CategoryTopLevel.Security:
        return resultOk<CategoryTopLevelUsecase>('Security')
      default:
        return resultError('Unhandled category type')
    }
  }

  private categoryTopLevelFromUsecase(category: CategoryTopLevelUsecase): CategoryTopLevel {
    switch (category) {
      case 'Health':
        return CategoryTopLevel.Health
      case 'Wealth':
        return CategoryTopLevel.Wealth
      case 'Security':
        return CategoryTopLevel.Security
    }
  }

  private impactFromUsecase(impact: ImpactUsecase): Impact {
    switch (impact) {
      case 'High':
        return Impact.High
      case 'Normal':
        return Impact.Normal
    }
  }

  private likelihoodFromUsecase(likelihood: LikelihoodUsecease): Likelihood {
    switch (likelihood) {
      case 'High':
        return Likelihood.High
      case 'Normal':
        return Likelihood.Normal
    }
  }

  private riskTypeFromUsecase(type: RiskTypeUsecase): RiskType {
    switch (type) {
      case 'Condition':
        return RiskType.Condition
      case 'Goal':
        return RiskType.Goal
      case 'Risk':
        return RiskType.Risk
    }
  }

  // ignore code coverage for risks until they are refactored
  /* istanbul ignore next */
  public riskFromUsecase(risk: RiskUsecase): Result<Risk> {
    let parent: Risk | undefined
    if (risk.parent) {
      const result = this.riskFromUsecase(risk.parent)
      if (!result.ok) {
        return result
      }
      parent = result.value
    }

    return resultOk({
      id: risk.id,
      category: this.categoryTopLevelFromUsecase(risk.category),
      impact: this.impactFromUsecase(risk.impact),
      likelihood: this.likelihoodFromUsecase(risk.likelihood),
      name: risk.name,
      notes: this.#mdxTranspiler(risk.notes),
      type: this.riskTypeFromUsecase(risk.type),
      shortDescription: risk.shortDescription,
      updated: risk.updated,
      parent,
    })
  }

  // ignore code coverage for risks until they are refactored
  /* istanbul ignore next */
  public risksFromUsecase(risks: RiskUsecase[]): Results<Risk> {
    return results(risks.map((risk) => this.riskFromUsecase(risk)))
  }

  public updatedFromUsecase(updated: UpdatedUsecase[]): Results<Category | Risk> {
    const mapped: Result<Category | Risk>[] = updated.map((u) => {
      if (isUpdatedCategoryUsecase(u)) {
        return resultOk(this.categoryFromUsecase(u))
      } else if (isUpdatedRiskUsecase(u)) {
        return this.riskFromUsecase(u)
      } else {
        return resultError<Category | Risk>('Unhandled Updated usecase type')
      }
    })
    return results(mapped)
  }

  public isUpdatedCategory(u: Category | Risk): u is Category {
    return 'children' in u && 'path' in u
  }
  public isUpdatedRisk(u: Category | Risk): u is Risk {
    return 'category' in u && 'impact' in u && 'likelihood' in u && 'type' in u
  }
}

export type { GraphMapper, MdxTranspiler }
export { newMapper }
