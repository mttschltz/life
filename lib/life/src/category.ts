import { IsArray, IsDate, IsOptional, IsString, MinLength, validateSync } from 'class-validator'
import { Result, resultError, resultOk } from '@util/result'
import { Update } from './update'

// TODO: children can also be risks (Concerns)

interface Category extends Update {
  id: string
  path: string
  name: string
  description?: string
  children: Category[]
  parent?: Category
}

type CreateDetails = Pick<
  Category,
  'children' | 'description' | 'name' | 'parent' | 'path' | 'shortDescription' | 'updated'
>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isCategory(category: any): category is Category {
  return (
    !!category &&
    /* eslint-disable @typescript-eslint/no-unsafe-member-access */
    category.id !== undefined &&
    category.path !== undefined &&
    category.name !== undefined &&
    category.children instanceof Array
    /* eslint-enable @typescript-eslint/no-unsafe-member-access */
  )
}

function newCategory(id: string, details: CreateDetails): Result<Category> {
  // Parent
  if (typeof details.parent !== 'undefined' && !isCategory(details.parent)) {
    return resultError('parent must be a Category')
  }

  // Children
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  if (!((details.children as any) instanceof Array) || details.children.some((c) => !isCategory(c))) {
    return resultError('children must be an array of Categorys')
  }
  const category = new CategoryImpl(details, id)

  // Primatives
  const errors = validateSync(category)
  if (errors.length > 0) {
    const constraints = errors[0].constraints
    if (!constraints) {
      return resultError('Validation failed')
    }
    return resultError(Object.values(constraints)[0])
  }

  return resultOk(category)
}

class CategoryImpl implements Category {
  /* eslint-disable @typescript-eslint/explicit-member-accessibility */
  #id: string

  #path: string
  #name: string
  #description?: string
  #shortDescription: string
  #parent?: Category
  #children: Category[]
  #updated: Date
  /* eslint-enable @typescript-eslint/explicit-member-accessibility */

  public constructor(details: CreateDetails, id: string) {
    this.#id = id

    this.#path = details.path
    this.#name = details.name
    this.#description = details.description
    this.#shortDescription = details.shortDescription
    this.#parent = details.parent
    this.#children = details.children
    this.#updated = details.updated
  }

  @MinLength(1)
  @IsString()
  public get id(): string {
    return this.#id
  }

  // Details

  @MinLength(1)
  @IsString()
  public get path(): string {
    return this.#path
  }

  @MinLength(2)
  @IsString()
  public get name(): string {
    return this.#name
  }

  @MinLength(1)
  @IsString()
  @IsOptional()
  public get description(): string | undefined {
    return this.#description
  }

  @MinLength(1)
  @IsString()
  public get shortDescription(): string {
    return this.#shortDescription
  }

  @IsOptional()
  public get parent(): Category | undefined {
    return this.#parent
  }

  @IsOptional()
  @IsArray()
  public get children(): Category[] {
    return this.#children
  }

  @IsDate()
  public get updated(): Date {
    return this.#updated
  }
}

export type { Category, CreateDetails }
export { newCategory }
