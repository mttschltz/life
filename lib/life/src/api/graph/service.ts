import { ApolloError, Config } from 'apollo-server'
import type { ListRisksCriteria } from '@life/usecase/listRisks'
import { Category as GraphCategory } from '@life/__generated__/graphql'
import type { QueryRisksArgs, RequireFields, Resolvers, Risk as GraphRisk } from '@life/__generated__/graphql'
import * as typeDefs from '@life/api/graph/schema.graphql'
import { InteractorFactory } from '@life/api/interactorFactory'
import { GraphMapper } from '@life/api/graph/mapper'
import { Logger } from '@util/logger'
import { ResultError } from '@util/result'
import { Impact, Likelihood, RiskType } from '@life/risk'

class GraphService {
  /* eslint-disable @typescript-eslint/explicit-member-accessibility */
  #factory: InteractorFactory
  #mapper: GraphMapper
  #logger: Logger
  /* eslint-enable @typescript-eslint/explicit-member-accessibility */

  public constructor(factory: InteractorFactory, mapper: GraphMapper, logger: Logger) {
    this.#factory = factory
    this.#mapper = mapper
    this.#logger = logger
  }

  // We may want to use Apollo dataSources at some point to leverage caching, deduplication and error handling.
  // They would probably need to call the usecases, rather than JsonRepo, as they have the relevant logic.
  // But for now, the benefits don't justify the overhead of adding another layer between resolvers and usecases.
  public resolvers(): Resolvers {
    return {
      Mutation: {
        createRisk: async (_, { input }): Promise<GraphRisk> => {
          // TODO: Instead of mapping each enum, etc individually, map CreateRiskInput
          // to CreateRiskRequest. This should make testing easier.
          const categoryResult = this.#mapper.toCategory(input.category)
          if (!categoryResult.ok) {
            this.#logger.result(categoryResult)
            throw this.resultError(categoryResult)
          }

          const riskResult = await this.#factory.createRiskInteractor().createRisk({
            category: categoryResult.value,
            name: input.name,
            parentId: input.parentId ? input.parentId : undefined,
            impact: Impact.Normal,
            likelihood: Likelihood.Normal,
            type: RiskType.Condition,
            uriPart: input.uriPart,
          })
          if (!riskResult.ok) {
            this.#logger.result(riskResult)
            throw this.resultError(riskResult)
          }

          const mappingResult = this.#mapper.fromRisk(riskResult.value)
          if (!mappingResult.ok) {
            this.#logger.result(mappingResult)
            throw this.resultError(mappingResult)
          }
          return mappingResult.value
        },
      },
      Risk: {
        parent: async (risk): Promise<GraphRisk | null> => {
          const parentResult = await this.#factory.fetchRiskParentInteractor().fetchRiskParent(risk.id)
          if (!parentResult.ok) {
            this.#logger.result(parentResult)
            throw this.resultError(parentResult)
          }

          const parent = parentResult.value
          if (!parent) {
            return null
          }

          const mappingResult = this.#mapper.fromRisk(parent)
          if (!mappingResult.ok) {
            this.#logger.result(mappingResult)
            throw this.resultError(mappingResult)
          }

          return mappingResult.value
        },
        children: async (risk): Promise<GraphRisk[]> => {
          const childrenResult = await this.#factory.fetchRiskChildrenInteractor().fetchRiskChildren(risk.id)
          if (!childrenResult.ok) {
            this.#logger.result(childrenResult)
            throw this.resultError(childrenResult)
          }

          const mappingResults = this.#mapper.risks(childrenResult.value)
          const errorResult = mappingResults.firstErrorResult
          if (errorResult) {
            this.#logger.result(errorResult)
            throw this.resultError(errorResult)
          }

          return mappingResults.okValues
        },
      },
      Query: {
        risks: async (_, args: RequireFields<QueryRisksArgs, never>): Promise<GraphRisk[]> => {
          const criteria: ListRisksCriteria = {}
          switch (args.category) {
            case GraphCategory.Health:
              criteria.category = 'health'
              break
            case GraphCategory.Wealth:
              criteria.category = 'wealth'
              break
            case GraphCategory.Security:
              criteria.category = 'security'
              break
          }

          const result = await this.#factory.listRisksInteractor().listRisks(criteria)
          if (!result.ok) {
            this.#logger.result(result)
            throw this.resultError(result)
          }

          const mappingResults = this.#mapper.risks(result.value)
          const errorResult = mappingResults.firstErrorResult
          if (errorResult) {
            this.#logger.result(errorResult)
            throw this.resultError(errorResult)
          }

          return mappingResults.okValues
        },
      },
    }
  }

  public typeDefs(): Config['typeDefs'] {
    return typeDefs
  }

  private resultError(result: ResultError): ApolloError {
    const error = result.error

    if (error?.stack) {
      return new ApolloError(result.message || 'Unknown error', undefined, {
        exception: {
          stacktrace: error.stack,
        },
      })
    }

    return new ApolloError(result.message || 'Unknown error')
  }
}

export { GraphService }
