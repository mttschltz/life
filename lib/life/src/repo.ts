import { Result, Results } from 'util/result'
import { CategoryTopLevel, Risk } from '@life/risk'
import { Category } from '@life/category'

interface RiskRepo {
  createRisk: (risk: Risk) => Promise<Result<void>>
  fetchRisk: (id: string) => Promise<Result<Risk>>
  fetchRiskChildren: (id: string) => Promise<Result<Risk[]>>
  fetchRiskParent: (id: string) => Promise<Result<Risk | undefined>>
  listRisks: (category: CategoryTopLevel | undefined, includeDescendents: boolean) => Promise<Result<Risk[]>>
}

interface CategoryRepo {
  fetchCategory: (id: string) => Promise<Result<Category>>
  fetchChildren: (id: string) => Promise<Results<Category>>
  fetchParent: (childId: string) => Promise<Result<Category | undefined>>
  list: (criteria: { includeChildren: boolean }) => Promise<Results<Category>>
}

export type { CategoryRepo, RiskRepo }
