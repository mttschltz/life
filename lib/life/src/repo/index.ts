import { Result } from 'lib/util'
import { Risk } from 'life/src'

export interface Repo {
  createRisk: (risk: Risk) => Result<void>
  fetchRisk: (id: string) => Result<Risk>
}
