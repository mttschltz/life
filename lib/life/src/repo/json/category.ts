// eslint-disable-next-line eslint-comments/disable-enable-pair
/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import { Result, resultError, resultOk, Results, results, resultsError, resultsErrorResult } from '@util/result'
import { CategoryRepo, CategoryRepo as CategoryRepoDomain } from '@life/repo'
import { CategoryJson, CategoryMapper } from '@life/repo/json/mapper'
import { Category } from '@life/category'
import { JsonStore } from './service'

type CategoryRepoJson = CategoryRepoDomain

function newCategoryRepoJson(store: JsonStore, mapper: CategoryMapper): CategoryRepoJson {
  return new CategoryRepoJsonImpl(store, mapper)
}

class CategoryRepoJsonImpl implements CategoryRepoJson {
  /* eslint-disable @typescript-eslint/explicit-member-accessibility */
  #store: JsonStore
  #mapper: CategoryMapper
  /* eslint-enable @typescript-eslint/explicit-member-accessibility */

  public constructor(store: JsonStore, mapper: CategoryMapper) {
    this.#store = store
    this.#mapper = mapper
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async fetch(id: string): Promise<Result<Category>> {
    const jsonCategory = this.#store.category[id]
    if (!jsonCategory) return resultError(`Could not find category '${id}'`)

    let parent
    if (jsonCategory.parentId) {
      const parentJson = this.#store.category[jsonCategory.parentId]
      if (!parentJson) {
        return resultError(`Could not find parent category '${jsonCategory.parentId}'`)
      }
      // We're ignoring parent and children here for simplicity, but may need to implement this later
      // in some manner that doesn't result in recursively loading parents/children.
      const parentMapResult = this.#mapper.fromJson(parentJson, undefined, [])
      if (!parentMapResult.ok) {
        return parentMapResult
      }
      parent = parentMapResult.value
    }

    let children: Category[] = []
    if (jsonCategory.children.length) {
      const childrenFetched: { [x: string]: CategoryJson }[] = jsonCategory.children.map((childId) => ({
        [childId]: this.#store.category[childId],
      }))
      const missingChildIds = childrenFetched
        .filter((c) => !Object.entries(c)[0][1])
        .map((c) => Object.entries(c)[0][0])
      if (missingChildIds.length) {
        return resultError(`Could not find child category '${missingChildIds[0]}'`)
      }
      const childrenMapResults = results(
        // We're ignoring parent and children here for simplicity, but may need to implement this later
        // in some manner that doesn't result in recursively loading parents/children.
        childrenFetched.map((json) => this.#mapper.fromJson(Object.values(json)[0], undefined, [])),
      )
      if (childrenMapResults.firstErrorResult) {
        return childrenMapResults.firstErrorResult
      }
      children = childrenMapResults.okValues
    }

    const categoryResult = this.#mapper.fromJson(jsonCategory, parent, children)
    if (!categoryResult.ok) {
      return categoryResult
    }

    return resultOk(categoryResult.value)
  }

  public async fetchParent(childId: string): Promise<Result<Category | undefined>> {
    const jsonCategory = this.#store.category[childId]
    if (!jsonCategory) return resultError(`Could not find category '${childId}'`)

    if (!jsonCategory.parentId) {
      return resultOk(undefined)
    }

    return this.fetch(jsonCategory.parentId)
  }

  public async fetchChildren(id: string): Promise<Results<Category>> {
    const jsonCategory = this.#store.category[id]
    if (!jsonCategory) return resultsError(`Could not find category '${id}'`)

    const childrenResults = await Promise.all(
      jsonCategory.children.map(async (childId) => {
        return this.fetch(childId)
      }),
    )
    return results(childrenResults)
  }

  public async list(criteria: Parameters<CategoryRepo['list']>[0]): Promise<Results<Category>> {
    const jsonCategories = Object.values(this.#store.category)
    const categoryResults = []
    for (const jsonCategory of jsonCategories) {
      if (criteria.onlyRoot && jsonCategory.parentId) {
        continue
      }

      const riskResult = await this.fetch(jsonCategory.id)
      if (!riskResult.ok) {
        return resultsErrorResult(riskResult)
      }

      categoryResults.push(riskResult)
    }
    return results(categoryResults)
  }
}

export type { CategoryRepoJson }
export { newCategoryRepoJson }
