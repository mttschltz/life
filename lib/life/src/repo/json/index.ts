import { Risk } from '@life'
import { Result } from '@util'
import { Category } from '@life/risk'
import { RiskRepo } from '@life/repo'
import { RiskJson, RiskMapper } from './mapper'

export interface Json {
  risk: {
    [key: string]: RiskJson
  }
}

export class JsonRepo implements RiskRepo {
  #json: Json
  #mapper: RiskMapper

  constructor(json: Partial<Json>, mapper: RiskMapper) {
    const risk = json.risk || {}
    this.#json = {
      risk,
    }
    this.#mapper = mapper
  }

  async createRisk(risk: Risk): Promise<Result<void>> {
    if (this.#json.risk[risk.id]) {
      return Result.error(`Risk with id '${risk.id}' already exists`)
    }

    if (risk.parent) {
      if (!this.#json.risk[risk.parent.id]) {
        return Result.error(`Could not find parent with id ${risk.parent.id} for risk with id ${risk.id}`)
      }
    }

    this.#json.risk[risk.id] = this.#mapper.toJson(risk)
    return Result.success(undefined)
  }

  async fetchRisk(id: string): Promise<Result<Risk>> {
    const jsonRisk = this.#json.risk[id]
    if (!jsonRisk) return Result.error(`Could not find risk ${id}`)

    let parent
    if (jsonRisk.parentId) {
      const parentResult = await this.fetchRisk(jsonRisk.parentId)
      if (!parentResult.isSuccess()) {
        return parentResult
      }
      parent = parentResult.getValue()
    }

    const riskResult = this.#mapper.fromJson(jsonRisk, parent)
    if (!riskResult.isSuccess()) {
      return riskResult
    }

    return Result.success(riskResult.getValue())
  }

  async fetchRiskParent(id: string): Promise<Result<Risk | undefined>> {
    const jsonRisk = this.#json.risk[id]
    if (!jsonRisk) return Result.error(`Could not find risk ${id}`)

    if (!jsonRisk.parentId) {
      return Result.success(undefined)
    }

    const parentRiskResult = await this.fetchRisk(jsonRisk.parentId)
    if (!parentRiskResult.isSuccess()) {
      return parentRiskResult
    }

    return Result.success(parentRiskResult.getValue())
  }

  async fetchRiskChildren(id: string): Promise<Result<Risk[]>> {
    const jsonRisks = Object.values(this.#json.risk)
    const riskChildren = []
    for (const jsonRisk of jsonRisks) {
      if (jsonRisk.parentId !== id) {
        continue
      }

      const childResult = await this.fetchRisk(jsonRisk.id)
      if (!childResult.isSuccess()) {
        return Result.errorFrom(childResult)
      }

      riskChildren.push(childResult.getValue())
    }
    return Result.success(riskChildren)
  }

  async listRisks(category: Category | undefined, includeDescendents: boolean): Promise<Result<Risk[]>> {
    const jsonRisks = Object.values(this.#json.risk)
    const risks = []
    for (let i = 0; i < jsonRisks.length; i++) {
      const jsonRisk = jsonRisks[i]
      if (category && jsonRisk.category !== category) {
        continue
      }

      if (!includeDescendents && jsonRisk.parentId) {
        continue
      }

      const riskResult = await this.fetchRisk(jsonRisk.id)
      if (!riskResult.isSuccess()) {
        return Result.errorFrom(riskResult)
      }

      risks.push(riskResult.getValue())
    }
    return Result.success(risks)
  }
}
