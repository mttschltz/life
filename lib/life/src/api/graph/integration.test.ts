import { JsonStore } from '@life/repo/json/service'
import { RiskMapper as RiskJsonMapper, newCategoryMapper } from '@life/repo/json/mapper'
import { ApolloServer } from 'apollo-server'
import { GraphService } from './service'
import { newMapper } from './mapper'
import {
  newCategoryInteractorFactory,
  newRiskInteractorFactory,
  newUpdatedInteractorFactory,
} from '@life/api/interactorFactory'
import { newLogger } from '@helper/logger'
import { transpile } from '@helper/mdx'
import { ApolloServerPluginLandingPageLocalDefault } from 'apollo-server-core'
import { gql } from '@apollo/client/core'
import { CategoryRepoJson, newCategoryRepoJson } from '@life/repo/json/category'
import { newRiskRepoJson, RiskRepoJson } from '@life/repo/json/risk'
import { UpdatedRepo as UpdatedRepoJson } from '@life/repo'
import { newUpdatedRepoJson } from '@life/repo/json/updated'
import { Impact, CategoryTopLevel, Likelihood, RiskType } from '@life/risk'

const jsonStore: JsonStore = {
  category: {
    1: {
      id: '1',
      name: 'root category 1',
      slug: 'root-category-1',
      previousSlugs: ['old-root-category-1'],
      path: '/root-category-1',
      previousPaths: ['/old-root-category-1'],
      description: 'root category 1 with child',
      shortDescription: 'rc1 with child',
      children: ['3'],
      updated: new Date('2021-12-01'),
    },
    2: {
      id: '2',
      name: 'root category 2',
      slug: 'root-category-2',
      previousSlugs: [],
      path: '/root-category-2',
      previousPaths: [],
      description: 'root category 2 no children',
      shortDescription: 'rc2 with child',
      children: [],
      updated: new Date('2021-12-02'),
    },
    3: {
      id: '3',
      name: 'child category 3',
      slug: 'child-category-3',
      previousSlugs: ['old-child-category-3'],
      path: '/child-category-3',
      previousPaths: ['/old-child-category-3'],
      description: 'child category 3 with child',
      shortDescription: 'cc3 with child',
      children: ['4'],
      parentId: '1',
      updated: new Date('2021-12-03'),
    },
    4: {
      id: '4',
      name: 'grandchild category 4',
      slug: 'grandchild-category-4',
      previousSlugs: [],
      path: '/grandchild-category-4',
      previousPaths: [],
      description: 'grandchild category 4 with child',
      shortDescription: 'gc4 with child',
      children: ['5'],
      parentId: '3',
      updated: new Date('2021-12-04'),
    },
    5: {
      id: '5',
      name: 'greatgrandchild category 5',
      slug: 'greatgrandchild-category-5',
      previousSlugs: [],
      path: '/greatgrandchild-category-5',
      previousPaths: [],
      description: 'greatgrandchild category 5 no children',
      shortDescription: 'ggc4 with child',
      children: [],
      parentId: '4',
      updated: new Date('2021-12-05'),
    },
  },
  risk: {
    1: {
      id: '1',
      category: CategoryTopLevel.Health,
      impact: Impact.High,
      likelihood: Likelihood.High,
      name: 'risk name 1',
      shortDescription: 'risk short description 1',
      type: RiskType.Condition,
      updated: new Date('2021-12-03'),
    },
    2: {
      id: '2',
      category: CategoryTopLevel.Health,
      impact: Impact.High,
      likelihood: Likelihood.High,
      name: 'risk name 2',
      shortDescription: 'risk short description 2',
      type: RiskType.Condition,
      updated: new Date('2021-12-04'),
      notes: 'risk notes 2',
      parentId: '1',
    },
  },
}

describe('GraphServiceIntegration', () => {
  let categoryRepo: CategoryRepoJson
  let riskRepo: RiskRepoJson
  let updatedRepo: UpdatedRepoJson
  let graphService: GraphService
  let server: ApolloServer
  beforeAll(async () => {
    categoryRepo = newCategoryRepoJson(jsonStore, newCategoryMapper())
    riskRepo = newRiskRepoJson(jsonStore, new RiskJsonMapper())
    updatedRepo = newUpdatedRepoJson(categoryRepo, riskRepo)

    graphService = new GraphService(
      {
        category: newCategoryInteractorFactory(categoryRepo),
        risk: newRiskInteractorFactory(riskRepo),
        updated: newUpdatedInteractorFactory(updatedRepo),
      },
      newMapper(transpile),
      newLogger(),
    )

    server = new ApolloServer({
      resolvers: graphService.resolvers(),
      typeDefs: graphService.typeDefs(),
      introspection: false,
      plugins: [ApolloServerPluginLandingPageLocalDefault()],
      mockEntireSchema: false,
    })

    await server.listen(4001)
  })
  afterAll(async () => {
    await server.stop()
  })

  describe('categories', () => {
    describe('Given a query for children and their parent', () => {
      const categoriesQuery = gql`
        query getCategories {
          categories {
            id
            slug
            previousSlugs
            name
            description
            shortDescription
            children {
              id
              slug
              previousSlugs
              name
              description
              shortDescription
              parent {
                id
                slug
                previousSlugs
                name
                description
                shortDescription
                updated
              }
              updated
            }
            updated
          }
        }
      `
      test('Then 2 root categories and 1 child with parent are returned', async () => {
        const response = await server.executeOperation({
          query: categoriesQuery,
        })
        expect(response).toMatchSnapshot()
      })
    })
  })
  describe('updated', () => {
    describe('Given a query with type-specific data', () => {
      const categoriesQuery = gql`
        query getUpdated {
          updated {
            id
            name
            shortDescription
            updated
            ... on Category {
              slug
              previousSlugs
              description
              parent {
                id
                slug
                previousSlugs
                name
                description
                shortDescription
                updated
              }
              children {
                id
                slug
                previousSlugs
                name
                description
                shortDescription
                updated
                parent {
                  id
                  slug
                  previousSlugs
                  name
                  description
                  shortDescription
                  updated
                }
              }
            }
            ... on Risk {
              category
              impact
              likelihood
              notes
              type
              parent {
                id
                category
                impact
                likelihood
                name
                notes
                type
                shortDescription
                updated
              }
            }
          }
        }
      `
      test('Then all categories and risks are returned with most recent updated first', async () => {
        const response = await server.executeOperation({
          query: categoriesQuery,
        })
        expect(response).toMatchSnapshot()
      })
    })
  })
})
