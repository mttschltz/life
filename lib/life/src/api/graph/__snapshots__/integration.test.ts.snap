// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphServiceIntegration categories Given a query for children and their parent Then 2 root categories and 1 child with parent are returned 1`] = `
Object {
  "data": Object {
    "categories": Array [
      Object {
        "children": Array [
          Object {
            "description": "child category 3 with child",
            "id": "3",
            "name": "child category 3",
            "parent": Object {
              "description": "root category 1 with child",
              "id": "1",
              "name": "root category 1",
              "path": "/root-category-1",
              "shortDescription": "rc1 with child",
              "updated": 2021-12-01T00:00:00.000Z,
            },
            "path": "/child-category-3",
            "shortDescription": "cc3 with child",
            "updated": 2021-12-03T00:00:00.000Z,
          },
        ],
        "description": "root category 1 with child",
        "id": "1",
        "name": "root category 1",
        "path": "/root-category-1",
        "shortDescription": "rc1 with child",
        "updated": 2021-12-01T00:00:00.000Z,
      },
      Object {
        "children": Array [],
        "description": "root category 2 no children",
        "id": "2",
        "name": "root category 2",
        "path": "/root-category-2",
        "shortDescription": "rc2 with child",
        "updated": 2021-12-02T00:00:00.000Z,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`GraphServiceIntegration updated Given a query with type-specific data Then 2 root categories and 2 risks are returned with most recent updated first 1`] = `
Object {
  "data": Object {
    "updated": Array [
      Object {
        "category": "HEALTH",
        "id": "2",
        "impact": "HIGH",
        "likelihood": "HIGH",
        "name": "risk name 2",
        "notes": "var _excluded = [\\"components\\"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/* @jsxRuntime classic */

/* @jsx mdx */
var layoutProps = {};
var MDXLayout = \\"wrapper\\";
return function MDXContent(_ref) {
  var components = _ref.components,
      props = _objectWithoutProperties(_ref, _excluded);

  return mdx(MDXLayout, _extends({}, layoutProps, props, {
    components: components,
    mdxType: \\"MDXLayout\\"
  }), mdx(\\"p\\", null, \\"risk notes 2\\"));
}
;
MDXContent.isMDXComponent = true;",
        "parent": Object {
          "category": "HEALTH",
          "id": "1",
          "impact": "HIGH",
          "likelihood": "HIGH",
          "name": "risk name 1",
          "notes": "Hello World",
          "shortDescription": "risk short description 1",
          "type": "CONDITION",
          "updated": 2021-12-03T00:00:00.000Z,
        },
        "shortDescription": "risk short description 2",
        "type": "CONDITION",
        "updated": 2021-12-04T00:00:00.000Z,
      },
      Object {
        "category": "HEALTH",
        "id": "1",
        "impact": "HIGH",
        "likelihood": "HIGH",
        "name": "risk name 1",
        "notes": "Hello World",
        "parent": null,
        "shortDescription": "risk short description 1",
        "type": "CONDITION",
        "updated": 2021-12-03T00:00:00.000Z,
      },
      Object {
        "children": Array [],
        "description": "root category 2 no children",
        "id": "2",
        "name": "root category 2",
        "parent": null,
        "path": "/root-category-2",
        "shortDescription": "rc2 with child",
        "updated": 2021-12-02T00:00:00.000Z,
      },
      Object {
        "children": Array [
          Object {
            "description": "child category 3 with child",
            "id": "3",
            "name": "child category 3",
            "parent": Object {
              "description": "root category 1 with child",
              "id": "1",
              "name": "root category 1",
              "path": "/root-category-1",
              "shortDescription": "rc1 with child",
              "updated": 2021-12-01T00:00:00.000Z,
            },
            "path": "/child-category-3",
            "shortDescription": "cc3 with child",
            "updated": 2021-12-03T00:00:00.000Z,
          },
        ],
        "description": "root category 1 with child",
        "id": "1",
        "name": "root category 1",
        "parent": null,
        "path": "/root-category-1",
        "shortDescription": "rc1 with child",
        "updated": 2021-12-01T00:00:00.000Z,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
